src/shared/lrc/lyrics.rs

```diff
@@ -1,114 +1,279 @@
-                            ));
-
-                        milis = match offset {
-                            Some(offset) if offset > 0 => {
-                                milis.saturating_sub(offset.unsigned_abs())
-                            }
-                            Some(offset) if offset < 0 => {
-                                milis.saturating_add(offset.unsigned_abs())
-                            }
-                            _ => milis,
-                        };
+            if !line_content.starts_with('[') {
+                continue;
+            }
 
-                        result.lines.push(LrcLine {
-                            time: Duration::from_millis(milis),
-                            content: line.trim().to_owned(),
-                        });
+            let mut remaining = &line_content[1..];
+            let mut tags = Vec::new();
+            let mut found_non_tag = false;
+            let mut lyrics_start = 0;
+            let mut offset_in_line = 1; // we skip the initial '[', so we want to include the first character after it in the offset
+
+            while !found_non_tag {
+                let mut bracket_count = 0;
+                let mut close_pos = None;
+                for (i, c) in remaining.char_indices() {
+                    match c {
+                        '[' => bracket_count += 1,
+                        ']' => {
+                            if bracket_count == 0 {
+                                close_pos = Some(i);
+                                break;
+                            }
+                            bracket_count -= 1;
+                        }
+                        _ => {}
                     }
                 }
-                Some(_) => {
-                    let (key, value) = meta_or_time
-                        .split_once(':')
-                        .with_context(|| format!("Invalid metadata line: '{meta_or_time}'"))?;
-                    match key.trim() {
-                        "offset" => offset = Some(value.trim().parse()?),
-                        "ti" => result.title = Some(value.trim().to_owned()),
-                        "ar" => result.artist = Some(value.trim().to_owned()),
-                        "al" => result.album = Some(value.trim().to_owned()),
-                        "au" => result.author = Some(value.trim().to_owned()),
-                        "length" => result.length = Some(parse_length(value.trim())?),
-                        _ => {}
+                let Some(close_pos) = close_pos else {
+                    break; // No closing bracket found
+                };
+                let tag_content = &remaining[..close_pos];
+                let is_timestamp = tag_content.chars().next().is_some_and(|c| c.is_numeric())
+                    && tag_content.contains(':');
+                let is_metadata = !is_timestamp && tag_content.contains(':');
+                if is_timestamp || is_metadata {
```

Comment:
```suggestion
                if is_timestamp && !is_metadata {
```

There shouldn't be a reason to push metadata to the tags vec at this point right? It gets ignored after.
Maybe we should also mention that metadata tags are ignored inside the actual lyrics in the file.

Reply:
Maybe not though, the current implementation also supports things like 
`[00:10.00][str:rst][00:12.00]repeated with mangled tag in the middle` which results in two repeated lyrics lines. The suggested change would break it.

But then again `[00:10.00][str][00:12.00]repeated with mangled tag in the middle` (without `:`) in the unexpected tag) already doesnt work with current impl. It resolves to a single line with `[str][00:12.00]repeated with mangled tag in the middle` as content.

Reply:
You're right, thanks for pointing that out, now I changed the behavior to be more conservative. The parser now **only processes consecutive timestamps at the start of a line**, stopping at the first non-timestamp tag (whether it's a valid metadata tag like `[str:rst]` or an invalid one like `[str]`).

For example:

```
[00:10.00][str:rst][00:12.00]lyrics
```

This creates a **single lyric line at `00:10.00`** with the content:

```
[str:rst][00:12.00]lyrics
```

It **does not attempt to parse** the later timestamp `[00:12.00]`.

While this might seem less "smart" when dealing with malformed files, i think it's a **safer choice**. It preserves expected behavior in normal cases like:

```
[00:10.00][Intro] Welcome
```

Where bracketed text like `[Intro]` should be treated as part of the lyrics, not misinterpreted as a timestamp or metadata.

If we try to implement something that handle those case we'll end up with a parser using complex heuristics to distinguish between metadata, invalid tags, and legitimate bracketed, which would be hard, time consuming, ambiguous and error-prone. This will also users to spot and clean up malformed timestamps in their files themselves when they are rendered.